www/docs/operations.mdx

```diff
@@ -0,0 +1,248 @@
+each other, and in fact, composition is so core to how operations work that
+_every operation in Effection eventually boils down to a combination of
+just three primitive operations_: `action()`, `resource()`, and `suspend()`.
+
+### Cancellation and Cleanup
+
+Perhaps the most critical difference between promises and operations is that
+once started, a promise will always run to completion no matter how long that
+takes. Unfortunately, this makes the guarantees of structured concurrency
+impossible to provide automatically. Operations, on the other hand, can be
+interrupted at any time which means that we never have to worry about them
+staying longer than they are absolutely needed.
+
+Let's consider a hypothetical sleep function implemented using `async/await`. It
+uses `setTimeout()` to resolve a promise once the sleep time has elapsed.
+
+```js
+async function sleep(duration) {
+  await new Promise(resolve => setTimeout(resolve, duration));
+}
+```
+
+We can now write our slow hello world program using this version:
+
+```js
+async function main() {
+  await sleep(1000);
+  console.log("Hello World!");
+}
+
+await main();
+```
+
+This logs `"Hello World!"` to the console after a delay of 1s, and so it works
+as expected when we run it... sort of.
+
+But what happens when you want to halt your sleep operation? Or more commonly,
+what happens when you no longer care about the outcome of your sleep operation?
+
+To understand why this might be, let's ask ourselves how long the program should
+take to execute in Node JS?
+
+```js
+async function sleep(milliseconds) {
+  await new Promise(resolve => setTimeout(resolve, milliseconds));
+}
+
+await Promise.race([sleep(10), sleep(10_000)]);
+```
+
+If you guessed 10 seconds, then you are technically correct! But if your
+intuition told you that the answer _should_ be 10 milliseconds, then
+congratulations, you are also correct!
+
+Our intuition tells us that the answer should be 10 milliseconds because when
+we race two promises against each other, and one of them wins, we don't care
+what happens to the other one. The result is now definitively known and so we
+can safely ignore all the other also-rans because they have absolutely no
+bearing on the outcome of the computation. But in this case, because the 10
+second sleep still has a pending `setTimeout()` callback installed, the run loop
+cannot exit, and even though we have absolutely nothing else to do, we're forced
+to wait around twiddling our thumbs until the timeout expires.
+
+By contrast, the equivalent code in Effection will exit immediately after 10ms
+as we expect.
+
+```js
+import { sleep, race, main } from "effection";
+
+await main(function*() {
+  yield* race([sleep(10), sleep(10_000)]);
+});
+```
+
+This is because every operation in Effection can be interrupted at any moment
+and as a result any cleanup code associated with that operation will be executed
+immediately.
```

Comment:
```md
By contrast, the equivalent code in Effection will exit immediately after 10ms as we expect because every operation in Effection can be interrupted at any time.
```